Yagro-Ltd/yagro-qa-wssyztf
# Drew bertoni

## Overview
Implementation of MS Playwright, including page object model and Playwright fixtures, and two test spec files for login page and crop page including visual and cross-browser coverage.

## Changes Made
- Uninstalled Cypress and removed related files/folders
- Installed Playwright and added tests

### Added
- /playwright folder
- /pages subfolder for page files (element locators and helper methods) containing page files for the login, dashboard, tracker and crop pages of the YAGRO test site
- /fixtures subfolder for fixtures file
- /tests subfolder for tests containing login and crop test specs and related subfolders for visual testing snapshots

### Modified
- Updated package.json and package-lock.json removing cypress and installing playwright v1.40.0
- Updated gitignore  to exclude folders used by Playwright to generate test reports and trace files/error videos
- Added new NPM aliases to package.json to trigger Playwright tests

## Testing Strategy
- **Test Cases:** 
- Login form: 
  - Visual test
  - Attempt to login without inputting email or password and check we remain on the login page
  - Attempt to login with email (but no password) and check we remain on the login page
  - Attempt to login with password (but no email) and check we remain on the login page
  - Attempt to login with invalid credentials and check we remain on the login page
  - Successfully login with a valid username and password and be taken to the dashboard page
- Crop progress page
  - Visual test
  - Select the 2021 year and check the main page values update correctly
  - Select the 2021 year and check the values in the Cost Details modal update correctly
  - Select the 2021 year and check that all of the main page elements load correctly
  - Open and dismiss the Crop Budgets modal using the Close button
  - Open and dismiss the Crop Budgets modal using the Cancel button
  - Open the Crop Budgets modal, toggle the total/per hectare switch and check that the Fixed Costs fuel values switch places
  - Open the Crop Budgets modal, then click OK and check we see a confirmation that no changes have been made
  - Update a value in the Crop Budgets modal and check the Unsaved Changes notification appears
  - Update a value in the Crop Budgets modal, then click cancel and check that no changes occur

- **Test Coverage:** 
login form, crop progress page

- **Testing Tools Used:**
MS Playwright

- **Manual Testing:**
N/A

## QA Checklist
Please ensure you have checked all the following items:

- [x] You have thoroughly tested your changes.
- [x] You have included relevant documentation (if applicable).

## Screenshots
Visual testing baseline screenshots generated on my Windows PC can be found in the subfolders inside /playwright/tests/

## Additional Notes
### Playwright info
**Running the tests**
After running `npm ci`, you should be able to use the npm aliases I added to package.json:

- `npm run playwright:snapshots` << IMPORTANT, please run this first to generate baseline snapshots for your host OS
- `npm run playwright:test` - run all tests
- `npm run playwright:ui` - run Playwright in UI mode

Full installation instructions are here: [https://playwright.dev/docs/intro](https://playwright.dev/docs/intro)
More information about the various additional test run options is available here: [https://playwright.dev/docs/running-tests](https://playwright.dev/docs/running-tests)

**Parallelisation**
Playwright will decide how many tests to run in parallel based on your available cores - this behaviour is controlled on line 22 of playwright.config.ts. However, my old PC laptop was struggling a bit and intermittently failing to load pages when running with 4 workers, so I had to restrict it to using 2 workers in local execution. Four or five parallel workers should be ok for a more up-to-date Macbook but you may need to play with this value to find a sweet spot.

**Additional browsers**
Tests are configured to run in desktop Chromium by default. You can run the tests in other browser engines by uncommenting those browser configurations in playwright.config.ts. The crop page test requires additional work to provide mobile coverage, so I've restricted this test to the desktop browsers only for now, but the login test should work fine in all configs including mobile.

### Notes about test logic
**Login form validaton**
I would normally add more form validation tests for a login form (eg. checking for malformed email addresses or passwords that are too short), but the form on the test site doesn't display any specific error messages, so I have skipped creating additional validation tests in the context of this exercise.

**Visual testing**
- As noted above, please ensure you update snapshots before running the tests, or visual tests will fail because baseline images don't exist for your browser and OS combination.
- I've set up visual testing to only execute in headless Chromium - this is mainly to show the capability and avoid generating tons of snapshots for all the different browser configurations, but there's also a known issue where snapshots generated in headless mode don't have scrollbars, and then fail if the test is run in headed browsers.
- I also masked the bar chart on the crop page in the visual test as its animation was causing failures - this doesn't seem to be a CSS animation, so I'd need to dig in a little bit to find a way to wait for it to complete. As a last resort, I could add a hard pause, but this is generally viewed as bad practice so I avoided it for now.

**Missing costs graph bug**
The Costs graph on the Crop page disappears if we change the selected year. This would be a blocking bug I'd raise in the normal way. I have added a test to cover possible future regression of this bug, so this test will fail until the issue is fixed.

### Linting
ESLint was being a huge pain on my Windows machine, so I decided to just use VSCode's inbuilt TS/JS linting instead of wasting precious time trying to get it running. Apologies for any minor formatting weirdness!
